### PR TITLE
Refactor proxy

### PR DESCRIPTION
**Pull request by andxu**
1. don't pollute user's proxy setting when user doesn't specify -Djava.net.useSystemProxies
2. don't pass http info here and there
3. read proxy from maven settings

### CODE REVIEW

File: azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/utils/ProxyUtils.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.math.NumberUtils;
+import org.apache.maven.execution.MavenExecutionRequest;
+import org.apache.maven.settings.Proxy;
+
+import java.util.List;
+
+public class ProxyUtils {
+    public static void initProxyManager(String httpProxyHost, String httpProxyPort, MavenExecutionRequest request) throws InvalidConfigurationException {
```
Comment (shenqianjin) [inline azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/utils/ProxyUtils.java:23]: how about to add it in ProxyManager together ?

Comment (andxu) [reply azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/utils/ProxyUtils.java:23]: this is only used for maven, while ProxyManager is shared for all toolkits.

**Review by shenqianjin — APPROVED** [review]
I have no more comment now...

File: azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/utils/ProxyUtils.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import java.util.List;
+import java.util.Objects;
+
+public class ProxyUtils {
+    public static void initProxyManager(MavenExecutionRequest request) {
+        final ProxyManager proxyManager = ProxyManager.getInstance();
+        if (Objects.isNull(proxyManager.getProxy())) {
+            final List<Proxy> mavenProxies = request.getProxies();
+            if (CollectionUtils.isNotEmpty(mavenProxies)) {
+                final org.apache.maven.settings.Proxy mavenProxy = mavenProxies.stream().filter(
```
Comment (Flanker32) [inline azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/utils/ProxyUtils.java:25]: no need to use full name here

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/AuthHelper.java [hunk 3]
```diff
@@ -56,11 +56,11 @@ public static void setupAzureEnvironment(AzureEnvironment env) {
         }
     }
 
-    public static void setupAuthProxy(String httpProxyHost, String httpProxyPort) {
+    public static void setupAuthProxy() {
+        final String httpProxyHost = ProxyManager.getInstance().getHttpProxyHost();
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/AuthHelper.java:60]: Can we `setupAuthProxy` in ProxyManager?

File: azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/proxy/ProxyManager.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            InetSocketAddress address = (InetSocketAddress) proxy.address();
+            return address.getPort();
+        }
+        return 0;
+    }
+
+    public void configure(@Nonnull String source, @Nonnull String httpProxyHost, @Nonnull Integer httpProxyPort) {
+        Preconditions.checkNotNull(httpProxyHost, "httpProxyHost must not be null.");
+        Preconditions.checkNotNull(httpProxyHost, "httpProxyPort must not be null.");
+        Preconditions.checkArgument(httpProxyPort.intValue() > 0, "httpProxyPort must be positive.");
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/proxy/ProxyManager.java:70]: also check port < 65535?

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/AuthHelper.java (hunk 3 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/AuthHelper.java]: ProxyManager in common-lib has no way to reference class in 'azure identity'

**Review by wangmingliang-ms — DISMISSED** [review]

File: azure-tools-common/src/main/java/com/microsoft/azure/common/applicationinsights/ApplicationInsightsManager.java [hunk 5]
```diff
@@ -13,7 +13,7 @@
 import com.microsoft.azure.management.applicationinsights.v2015_05_01.ApplicationType;
 import com.microsoft.azure.management.applicationinsights.v2015_05_01.implementation.InsightsManager;
 import com.microsoft.azure.management.resources.Provider;
-import com.microsoft.azure.tools.common.util.ProxyUtils;
+import com.microsoft.azure.toolkit.lib.common.proxy.ProxyManager;
```
Comment (wangmingliang-ms) [inline azure-tools-common/src/main/java/com/microsoft/azure/common/applicationinsights/ApplicationInsightsManager.java:16]: remove, the manager doesn't need to know proxy

File: azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/proxy/ProxyManager.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    public int getHttpProxyPort() {
+        if (Objects.nonNull(this.proxy) && proxy.address() instanceof InetSocketAddress) {
+            InetSocketAddress address = (InetSocketAddress) proxy.address();
+            return address.getPort();
+        }
+        return 0;
+    }
+
+    public void configure(@Nonnull String httpProxyHost, @Nonnull Integer httpProxyPort) {
+        Preconditions.checkNotNull(httpProxyHost, "httpProxyHost must not be null.");
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/proxy/ProxyManager.java:61]: Objects.requireNotNull

Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/proxy/ProxyManager.java:61]: actually there are no difference https://stackoverflow.com/questions/34646554/java-util-objects-requirenonnull-vs-preconditions-checknotnull

File: azure-tools-common/src/main/java/com/microsoft/azure/common/applicationinsights/ApplicationInsightsManager.java (hunk 5 above)
Comment (andxu) [reply azure-tools-common/src/main/java/com/microsoft/azure/common/applicationinsights/ApplicationInsightsManager.java]: fixed

**Review by wangmingliang-ms — DISMISSED** [review]

**Review by wangmingliang-ms — APPROVED** [review]

**Pull request merged**